samples/infiniteStreaming.js

```diff
@@ -0,0 +1,284 @@
+    audioInput = [];
+    // Initiate (Reinitiate) a recognize stream
+    recognizeStream = client
+      .streamingRecognize(request)
+      .on('error', err => {
+        if (err.code === 11) {
+          // uncomment the following code to use timeout error to restart
+          //restartStream();
+        } else {
+          console.error('API request error ' + err);
+        }
+      })
+      .on('data', speechCallback);
+
+    //restart stream when streamingLimit expires
+    setTimeout(restartStream, streamingLimit);
+  }
+
+  const speechCallback = stream => {
+    //convert API result end time from seconds + nanoseconds into milliseconds
+    resultEndTime =
+      stream.results[0].resultEndTime.seconds * 1000 +
+      Math.round(stream.results[0].resultEndTime.nanos / 1000000);
+
+    //calculate correct time based on offset from audio sent twice
+    const correctedTime =
+      resultEndTime - bridgingOffset + streamingLimit * restartCounter;
+
+    process.stdout.clearLine();
+    process.stdout.cursorTo(0);
+    let stdoutText = '';
+    if (stream.results[0] && stream.results[0].alternatives[0]) {
+      stdoutText =
+        correctedTime + ': ' + stream.results[0].alternatives[0].transcript;
+    }
+
+    if (stream.results[0].isFinal) {
+      process.stdout.write(chalk.green(`${stdoutText}\n`));
+
+      isFinalEndTime = resultEndTime;
+      lastTranscriptWasFinal = true;
+    } else {
+      //make sure transcript does not exceed console character length
+      if (stdoutText.length > process.stdout.columns) {
+        stdoutText =
+          stdoutText.substring(0, process.stdout.columns - 4) + '...';
+      }
+      process.stdout.write(chalk.red(`${stdoutText}`));
+
+      lastTranscriptWasFinal = false;
+    }
+  };
+
+  const audioInputStreamTransform = new Transform({
+    transform: (chunk, encoding, callback) => {
+      if (newStream && lastAudioInput.length !== 0) {
+        //approximate math to calculate time of chunks
+        const chunkTime = streamingLimit / lastAudioInput.length;
+        if (chunkTime !== 0) {
+          if (bridgingOffset < 0) {
+            bridgingOffset = 0;
+          }
+          if (bridgingOffset > finalRequestEndTime) {
+            bridgingOffset = finalRequestEndTime;
+          }
+          const chunksFromMS = Math.floor(
+            (finalRequestEndTime - bridgingOffset) / chunkTime
+          );
+          bridgingOffset = Math.floor(
+            (lastAudioInput.length - chunksFromMS) * chunkTime
+          );
+
+          for (let i = chunksFromMS; i < lastAudioInput.length; i++) {
+            recognizeStream.write(lastAudioInput[i]);
+          }
+        }
+        newStream = false;
+      }
+
+      audioInput.push(chunk);
+
+      if (recognizeStream) {
+        recognizeStream.write(chunk);
+      }
+
+      callback();
+    },
+  });
+
+  function restartStream() {
+    if (recognizeStream) {
+      recognizeStream.removeListener('data', speechCallback);
```

Comment:
I think `recognizeStream.removeAllListeners()` should do the trick here.

Reply:
I tried this out, and it resulted in more listeners being removed than intended. I think there are listeners connecting to the helpers.js file in the @google-cloud/speech client library.
Also, I'm currently suppressing a code 11 error on line 104, which is a timeout error that doesn't seem to reflect what is happening with the audio stream, but the error could be used to trigger the stream restart in the future... There may be a better way to close the stream in the helper file, now that there is new functionality. perhaps the helpers.js file is where this functionality should live.